stdlib/public/core/Array.swift

```diff
@@ -1865,3 +1865,168 @@ internal struct _ArrayAnyHashableBox<Element: Hashable>
+    public init(elements: [Element.TangentVector]) {
+      self.elements = elements
+    }
+
+    @inlinable
+    public static func + (lhs: TangentVector, rhs: TangentVector) -> TangentVector {
+      return TangentVector(elements: zipLongest(lhs.elements, rhs.elements).map {
+        ($0 ?? .zero) + ($1 ?? .zero)
+      })
+    }
+
+    @inlinable
+    public static func - (lhs: TangentVector, rhs: TangentVector) -> TangentVector {
+      return TangentVector(elements: zipLongest(lhs.elements, rhs.elements).map {
+        ($0 ?? .zero) - ($1 ?? .zero)
+      })
+    }
+
+    @inlinable
+    public func moved(along direction: TangentVector) -> [Element.TangentVector] {
+      return zip(elements, direction.elements).map { $0.moved(along: $1) }
+    }
+
+    @inlinable
+    public func tangentVector(from cotangent: CotangentVector) -> TangentVector {
+      let elements = zip(self.elements, cotangent.elements).map {
+        $0.tangentVector(from: $1)
+      }
+      return TangentVector(elements: elements)
+    }
+  }
+
+  public struct CotangentVector : AdditiveArithmetic & Differentiable {
+    public typealias TangentVector = Array<Element>.CotangentVector
+    public typealias CotangentVector = Array<Element>.TangentVector
+
+    public var elements: [Element.CotangentVector]
+
+    public static var zero: TangentVector {
+      return TangentVector(elements: [])
+    }
+    
+    public init(elements: [Element.CotangentVector]) {
+      self.elements = elements
+    }
+
+    @inlinable
+    public static func + (lhs: TangentVector, rhs: TangentVector) -> TangentVector {
+      return TangentVector(elements: zipLongest(lhs.elements, rhs.elements).map {
+        ($0 ?? .zero) + ($1 ?? .zero)
+      })
+    }
+
+    @inlinable
+    public static func - (lhs: TangentVector, rhs: TangentVector) -> TangentVector {
+      return TangentVector(elements: zipLongest(lhs.elements, rhs.elements).map {
+        ($0 ?? .zero) - ($1 ?? .zero)
+      })
+    }
+
+    @inlinable
+    public func moved(along direction: TangentVector) -> [Element.CotangentVector] {
+      return zip(elements, direction.elements).map { $0.moved(along: $1) }
+    }
+
+    @inlinable
+    public func tangentVector(from cotangent: CotangentVector) -> TangentVector {
+      let elements = zip(self.elements, cotangent.elements).map {
+        $0.tangentVector(from: $1)
+      }
+      return TangentVector(elements: elements)
+    }
+  }
+
+  public typealias AllDifferentiableVariables = [Element]
+
+  @inlinable
+  public func moved(along direction: TangentVector) -> [Element] {
+    return zip(self, direction.elements).map { $0.moved(along: $1) }
+  }
+
+  @inlinable
+  public func tangentVector(from cotangent: CotangentVector) -> TangentVector {
+    return TangentVector(elements: zip(self, cotangent.elements).map { $0.tangentVector(from: $1) })
+  }
+}
+
+// TODO: Remove once `zipLongest` is part of stdlib.
+public func zipLongest<Sequence1: Sequence, Sequence2: Sequence>(
```

Comment:
Helper functions in stdlib shouldn't be public and should be underscored. `@usableFromInline` makes you able to call it from a `@inlinable` context.
```suggestion
@usableFromInline
internal func _zipLongest<Sequence1: Sequence, Sequence2: Sequence>(
```